### PR TITLE
Implement for new ParameterCollection interface

### PR DESCRIPTION
**Pull request by xunzhang**
- [x] Implement get_fullname and get_namespace interfaces, add related tests.
- [x] Add interface for supporting m.add_parameters({10}, "a") and tests.
- [x] Add input parameter and add_subcollection verify.
- [x] Handle parameter copy logic: currently code treat copy as sharing objects with original ones while we might disable copy in the future.
- [x] Parameter retrieval interfaces implementation. By model, by name, etc.
- [x] Builder classes implementation and tests.

### CODE REVIEW

**Comment by xunzhang** [comment]
Will file remaining sub-tasks in following pull request.

- Parameter/lookupParameters level Serializer implementation, load/save functions.
- Python interface wrapper and tests.
- Update tutorial

File: dynet/cfsm-builder.h [hunk 1]
```diff
@@ -38,7 +38,12 @@ class StandardSoftmaxBuilder : public SoftmaxBuilder {
   expr::Expression neg_log_softmax(const expr::Expression& rep, unsigned wordidx);
   unsigned sample(const expr::Expression& rep);
   expr::Expression full_log_distribution(const expr::Expression& rep);
-
+  std::vector<ParameterStorage*> get_parameters() {
```
Comment (neubig) [inline dynet/cfsm-builder.h]: In this and the other classes in the commit, wouldn't be easier to save `local_model` as a member of the class, then when `get_parameters()` is called, call `local_model.get_parameters()`? The ability to do this simply is one of the major advantages of the `ParameterCollection` concept I think.

File: dynet/model.cc [hunk 2]
```diff
@@ -85,6 +83,11 @@ DYNET_SERIALIZE_COMMIT(ParameterStorage,
 DYNET_SERIALIZE_IMPL(ParameterStorage)
 #endif
 
+bool validParameter(const std::string & s) {
```
Comment (neubig) [inline dynet/model.cc]: This should be in snake case.

File: dynet/cfsm-builder.h (hunk 1 above)
Comment (xunzhang) [inline dynet/cfsm-builder.h]: Agree!

File: dynet/model.cc (hunk 2 above)
Comment (xunzhang) [inline dynet/model.cc]: fixed, thanks

**Comment by neubig** [comment]
Thanks! Looks good.

**Pull request merged**